src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java

```diff
@@ -119,4 +127,64 @@ public Stats deserialize(JsonElement jsonElement, Type type,
             }
         }
     }
+
+    /**
+     * Deserializer for Arguments class.
+     * Breaks the incoming arguments into args and kwargs parts
+     * and fills a new Arguments instance.
+     */
+    private class ArgumentsDeserializer implements JsonDeserializer<Arguments> {
+
+        private static final String KWARG_KEY = "__kwarg__";
+
+        @Override
+        public Arguments deserialize(JsonElement json, Type typeOfT,
+                JsonDeserializationContext context) throws JsonParseException {
+            Arguments result = new Arguments();
+
+            if (json != null && json.isJsonArray()) {
+                for (JsonElement jsonElement : json.getAsJsonArray()) {
+                    fillArgs(result, jsonElement);
+                }
+            }
+
+            return result;
+        }
+
+        private void fillArgs(Arguments result, JsonElement jsonElement) {
+            if (handleKwarg(jsonElement)) {
+                fillKwargsFromObject(result, jsonElement.getAsJsonObject());
+            } else {
+                result.getArgs().add(gson.fromJson(jsonElement, Object.class));
+            }
+        }
+
+        private void fillKwargsFromObject(Arguments result, JsonObject jsonObject) {
+            for (Map.Entry<String, JsonElement> kwItem : jsonObject.entrySet()) {
+                result.getKwargs().put(kwItem.getKey(),
+                        gson.fromJson(kwItem.getValue(),
+                        Object.class));
+            }
+        }
+
+        /**
+         * Determines whether an object should be handled as a positional or keyword warg.
+         * An element is handled as a keyword argument if it's an object and has __kwarg__
+         * flag set to true.
+         * As a side effect, this method removes a property marking the object as arg/kwarg
+         * as this is not needed further.
+         * @param jsonElement to be tested
+         * @return true if this element is a keyword argument
+         */
+        private boolean handleKwarg(JsonElement jsonElement) {
+            boolean kwarg = false;
+            if (jsonElement.isJsonObject()) {
+                JsonObject asJsonObject = jsonElement.getAsJsonObject();
+                kwarg = asJsonObject.has(KWARG_KEY)
+                        && Boolean.TRUE.equals(asJsonObject.get(KWARG_KEY).getAsBoolean());
```

Comment:
Hi,
* as for the `TRUE.equals` - it's there for null check reasons. This is handy when an evil user passes `__kwarg__: null`. We could also try to `getAsBoolean` and catch exceptions, but I think that's equally ugly :)
* handleKwarg stuff: if I put removing `__kwarg__` attribute to `fillKwargsFromObject`, it won't remove this attribute in case the passed arg is an object and has `__kwarg__` set to something else than `true`.
The question is: should we store `__kwarg__` in our representation of args/kwargs? I don't think so because this information is already contained - by placing attributes to `args`/`kwargs` attribute of `Arguments` class.